### PR TITLE
 [DAG] isSplatValue - node is a splat if all demanded elts have the same whole constant value

### PR DESCRIPTION
**Pull request by RKSimon**

### CODE REVIEW

File: llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll [hunk 1]
```diff
@@ -138,7 +141,10 @@ define void @program_1(ptr %dest, ptr %t0, <4 x float> %p0, <4 x float> %p1, <4
 ; X64-NEXT:    xorps %xmm0, %xmm0
 ; X64-NEXT:    movaps %xmm0, {{[-0-9]+}}(%r{{[sb]}}p) ## 16-byte Spill
 ; X64-NEXT:    movaps {{[-0-9]+}}(%r{{[sb]}}p), %xmm0 ## 16-byte Reload
-; X64-NEXT:    mulps %xmm0, %xmm0
+; X64-NEXT:    mulss {{\.?LCPI[0-9]+_[0-9]+}}(%rip), %xmm0
+; X64-NEXT:    movaps %xmm0, {{[-0-9]+}}(%r{{[sb]}}p) ## 16-byte Spill
+; X64-NEXT:    movaps {{[-0-9]+}}(%r{{[sb]}}p), %xmm0 ## 16-byte Reload
+; X64-NEXT:    shufps {{.*#+}} xmm0 = xmm0[0,0,0,0]
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll:147]: NOTE: This regression is addressed by #74304

Comment (phoebewang) [reply llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll:147]: If I read the IR correctly, it is a `zeroinitializer * zeroinitializer`. Why we still need a `mulss` here?

File: llvm/test/CodeGen/X86/var-permute-256.ll [hunk 2]
```diff
@@ -25,18 +25,20 @@ define <4 x i64> @var_shuffle_v4i64(<4 x i64> %v, <4 x i64> %indices) nounwind {
 ;
 ; AVX1-LABEL: var_shuffle_v4i64:
 ; AVX1:       # %bb.0:
-; AVX1-NEXT:    vperm2f128 {{.*#+}} ymm2 = ymm0[2,3,2,3]
-; AVX1-NEXT:    vpaddq %xmm1, %xmm1, %xmm3
-; AVX1-NEXT:    vextractf128 $1, %ymm1, %xmm1
+; AVX1-NEXT:    vextractf128 $1, %ymm1, %xmm2
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/var-permute-256.ll:28]: Didn't look into details, but the RHS has more instructions. Is it intended?

Comment (RKSimon) [reply llvm/test/CodeGen/X86/var-permute-256.ll:28]: Its replaced a ymm constant pool entry of `<4 x i64> <i64 2,i64 2,i64 2,i64 2>`, which is then split into 2 xmm folded loads in vpcmpgtq - into a single vmovddup. So increase by 1 instruction (extra line for comment), but replaced 2 xmm folded loads with a single broadcast load.

File: llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll (hunk 1 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll:147]: Various hasOneUse limits prevents the constant data ever being seen as a fp constant, preventing constant folding.

File: llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll [hunk 3]
```diff
@@ -54,7 +54,10 @@ define void @program_1(ptr %dest, ptr %t0, <4 x float> %p0, <4 x float> %p1, <4
 ; X32-NEXT:    xorps %xmm0, %xmm0
 ; X32-NEXT:    movaps %xmm0, {{[-0-9]+}}(%e{{[sb]}}p) ## 16-byte Spill
 ; X32-NEXT:    movaps {{[-0-9]+}}(%e{{[sb]}}p), %xmm0 ## 16-byte Reload
-; X32-NEXT:    mulps %xmm0, %xmm0
+; X32-NEXT:    mulss {{\.?LCPI[0-9]+_[0-9]+}}, %xmm0
```
[the remark this answers is not in the document]
Comment (RKSimon) [reply llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll:57]: Yeah, I missed this one - every so often I fix some files but there's always more :)

File: llvm/test/CodeGen/X86/var-permute-256.ll (hunk 2 above)
Comment (phoebewang) [reply llvm/test/CodeGen/X86/var-permute-256.ll:28]: Got it, thanks!

File: llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll (hunk 1 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/fold-pcmpeqd-2.ll:147]: Its getting narrowed from mulps to mulss by scalarizeBinOpOfSplats

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**